// y-quill/src/y-quill.js
/**
 * @module bindings/quill
 */

import { createMutex } from 'lib0/mutex.js'
import * as Y from 'yjs' // eslint-disable-line
import { Awareness } from 'y-protocols/awareness.js' // eslint-disable-line

/**
 * Removes the pending '\n's if it has no attributes.
 */
export const normQuillDelta = delta => {
    if (delta.length > 0) {
        const d = delta[delta.length - 1]
        const insert = d.insert
        if (d.attributes === undefined && insert !== undefined && insert.slice(-1) === '\n') {
            delta = delta.slice()
            let ins = insert.slice(0, -1)
            while (ins.slice(-1) === '\n') {
                ins = ins.slice(0, -1)
            }
            delta[delta.length - 1] = { insert: ins }
            if (ins.length === 0) {
                delta.pop()
            }
            return delta
        }
    }
    return delta
}

/**
 * @param {any} quillCursors
 */
const updateCursor = (quillCursors, aw, clientId, doc, type) => {
    try {
        if (aw && aw.cursor && clientId !== doc.clientID) {
            const user = aw.user || {}
            const color = user.color || '#ffa500'
            const name = user.name || `User: ${clientId}`
            quillCursors.createCursor(clientId.toString(), name, color)
            const anchor = Y.createAbsolutePositionFromRelativePosition(
                Y.createRelativePositionFromJSON(aw.cursor.anchor),
                doc
            )
            const head = Y.createAbsolutePositionFromRelativePosition(
                Y.createRelativePositionFromJSON(aw.cursor.head),
                doc
            )
            if (anchor && head && anchor.type === type) {
                quillCursors.moveCursor(clientId.toString(), { index: anchor.index, length: head.index - anchor.index })
            }
        } else {
            quillCursors.removeCursor(clientId.toString())
        }
    } catch (err) {
        console.error(err)
    }
}

// --- DEBUG LOGGING for format bleed investigation ---
const YQ_DEBUG = true
const _yqLog = (...args) => {
    if (YQ_DEBUG) console.log('[y-quill]', ...args)
}
const _yqWarn = (...args) => {
    if (YQ_DEBUG) console.warn('[y-quill]', ...args)
}
const _snippet = (text, maxLen = 30) => {
    if (typeof text !== 'string') return JSON.stringify(text)
    if (text.length <= maxLen) return JSON.stringify(text)
    return JSON.stringify(text.slice(0, maxLen) + '…')
}
const _fmtAttrs = attrs => {
    if (!attrs || Object.keys(attrs).length === 0) return '{}'
    const parts = []
    for (const k in attrs) {
        parts.push(`${k}: ${JSON.stringify(attrs[k])}`)
    }
    return `{ ${parts.join(', ')} }`
}
// --- END DEBUG ---

export class QuillBinding {
    /**
     * @param {Y.Text} type
     * @param {any} quill
     * @param {Awareness} [awareness]
     */
    constructor(type, quill, awareness) {
        const mux = createMutex()
        const doc = /** @type {Y.Doc} */ (type.doc)
        this.mux = mux
        this.type = type
        this.doc = doc
        this.quill = quill
        const quillCursors = quill.getModule('cursors') || null
        this.quillCursors = quillCursors
        // This object contains all attributes used in the quill instance
        this._negatedUsedFormats = {}
        this.awareness = awareness
        this._awarenessChange = ({ added, removed, updated }) => {
            const states = /** @type {Awareness} */ (awareness).getStates()
            added.forEach(id => {
                updateCursor(quillCursors, states.get(id), id, doc, type)
            })
            updated.forEach(id => {
                updateCursor(quillCursors, states.get(id), id, doc, type)
            })
            removed.forEach(id => {
                quillCursors.removeCursor(id.toString())
            })
        }
        this._typeObserver = event => {
            mux(() => {
                const eventDelta = event.delta
                _yqLog('--- _typeObserver (Yjs→Quill) ---')
                _yqLog('  _negatedUsedFormats:', JSON.stringify(this._negatedUsedFormats))
                _yqLog('  eventDelta (' + eventDelta.length + ' ops):')
                eventDelta.forEach((d, i) => {
                    if (d.insert !== undefined) {
                        _yqLog(`    [${i}] INSERT ${_snippet(d.insert)} attrs=${_fmtAttrs(d.attributes)}`)
                    } else if (d.retain !== undefined) {
                        _yqLog(`    [${i}] RETAIN ${d.retain} attrs=${_fmtAttrs(d.attributes)}`)
                    } else if (d.delete !== undefined) {
                        _yqLog(`    [${i}] DELETE ${d.delete}`)
                    }
                })
                // We always explicitly set attributes, otherwise concurrent edits may
                // result in quill assuming that a text insertion shall inherit existing
                // attributes.
                const delta = []
                for (let i = 0; i < eventDelta.length; i++) {
                    const d = eventDelta[i]
                    if (d.insert !== undefined) {
                        const mergedAttrs = Object.assign({}, this._negatedUsedFormats, d.attributes || {})
                        delta.push(
                            Object.assign({}, d, {
                                attributes: mergedAttrs,
                            })
                        )
                    } else {
                        delta.push(d)
                    }
                }
                _yqLog('  normalized delta sent to Quill:')
                delta.forEach((d, i) => {
                    if (d.insert !== undefined) {
                        _yqLog(`    [${i}] INSERT ${_snippet(d.insert)} attrs=${_fmtAttrs(d.attributes)}`)
                    } else if (d.retain !== undefined) {
                        _yqLog(`    [${i}] RETAIN ${d.retain} attrs=${_fmtAttrs(d.attributes)}`)
                    } else if (d.delete !== undefined) {
                        _yqLog(`    [${i}] DELETE ${d.delete}`)
                    }
                })
                quill.updateContents(delta, 'yjs')
            })
        }
        type.observe(this._typeObserver)
        this._quillObserver = (eventType, delta) => {
            if (delta && delta.ops) {
                // update content
                const ops = delta.ops
                _yqLog('--- _quillObserver (Quill→Yjs) --- eventType:', eventType)
                _yqLog('  incoming ops (' + ops.length + '):')
                ops.forEach((op, i) => {
                    if (op.insert !== undefined) {
                        _yqLog(`    [${i}] INSERT ${_snippet(op.insert)} attrs=${_fmtAttrs(op.attributes)}`)
                    } else if (op.retain !== undefined) {
                        _yqLog(`    [${i}] RETAIN ${op.retain} attrs=${_fmtAttrs(op.attributes)}`)
                    } else if (op.delete !== undefined) {
                        _yqLog(`    [${i}] DELETE ${op.delete}`)
                    }
                })
                ops.forEach(op => {
                    if (op.attributes !== undefined) {
                        for (let key in op.attributes) {
                            if (this._negatedUsedFormats[key] === undefined) {
                                this._negatedUsedFormats[key] = false
                                _yqLog('  new format key tracked:', key)
                            }
                        }
                    }
                })
                _yqLog('  _negatedUsedFormats:', JSON.stringify(this._negatedUsedFormats))
                mux(() => {
                    // FIX: Yjs applyDelta() doesn't properly handle null attributes for format removal.
                    // See https://github.com/yjs/yjs/issues/474
                    // We need to manually call type.format() for retain ops with null attributes,
                    // as applyDelta() doesn't reliably remove attributes when set to null.

                    let currentIndex = 0
                    const opsWithoutNullFormats = []
                    const formatOps = [] // Collect format removal operations to apply after

                    for (let i = 0; i < ops.length; i++) {
                        const op = ops[i]

                        if (op.retain !== undefined && op.attributes) {
                            // Check if any attributes are null (format removal)
                            const nullAttrs = {}
                            const nonNullAttrs = {}
                            let hasNullAttrs = false
                            let hasNonNullAttrs = false

                            for (const key in op.attributes) {
                                if (op.attributes[key] === null) {
                                    nullAttrs[key] = null
                                    hasNullAttrs = true
                                } else {
                                    nonNullAttrs[key] = op.attributes[key]
                                    hasNonNullAttrs = true
                                }
                            }

                            if (hasNullAttrs) {
                                // Store format removal operation to apply explicitly
                                formatOps.push({
                                    index: currentIndex,
                                    length: op.retain,
                                    attributes: nullAttrs,
                                })

                                // If there are also non-null attrs, keep them in the delta
                                if (hasNonNullAttrs) {
                                    opsWithoutNullFormats.push({
                                        retain: op.retain,
                                        attributes: nonNullAttrs,
                                    })
                                } else {
                                    // Just retain without attributes
                                    opsWithoutNullFormats.push({ retain: op.retain })
                                }
                            } else {
                                // No null attrs, keep as-is
                                opsWithoutNullFormats.push(op)
                            }
                            currentIndex += op.retain
                        } else if (op.retain !== undefined) {
                            opsWithoutNullFormats.push(op)
                            currentIndex += op.retain
                        } else if (op.insert !== undefined) {
                            opsWithoutNullFormats.push(op)
                            const insertLength = typeof op.insert === 'string' ? op.insert.length : 1
                            // Prevent Yjs attribute inheritance on inserts.
                            // When text is inserted adjacent to formatted text without explicit
                            // format attributes, Yjs may inherit formats from adjacent text
                            // (e.g. during CRDT merge). Explicitly clear missing formats
                            // using type.format() after applyDelta, same pattern as the
                            // retain-fix for Yjs issue #474.
                            const insertAttrs = op.attributes || {}
                            const clearAttrs = {}
                            let hasClearAttrs = false
                            for (const key in this._negatedUsedFormats) {
                                if (insertAttrs[key] === undefined) {
                                    clearAttrs[key] = null
                                    hasClearAttrs = true
                                }
                            }
                            if (hasClearAttrs) {
                                formatOps.push({
                                    index: currentIndex,
                                    length: insertLength,
                                    attributes: clearAttrs,
                                })
                            }
                            currentIndex += insertLength
                        } else if (op.delete !== undefined) {
                            opsWithoutNullFormats.push(op)
                            // delete doesn't advance the index in the same way
                        } else {
                            opsWithoutNullFormats.push(op)
                        }
                    }

                    if (formatOps.length > 0) {
                        _yqLog('  formatOps to apply after applyDelta:')
                        formatOps.forEach((fo, i) => {
                            _yqLog(`    [${i}] index=${fo.index} length=${fo.length} attrs=${_fmtAttrs(fo.attributes)}`)
                        })
                    }

                    // Apply the delta without the null format operations
                    if (opsWithoutNullFormats.length > 0) {
                        _yqLog('  applying delta to Yjs (' + opsWithoutNullFormats.length + ' ops)')
                        type.applyDelta(opsWithoutNullFormats)
                    }

                    // Now explicitly apply format removals using type.format()
                    // This ensures null attributes actually remove the formatting
                    for (const formatOp of formatOps) {
                        _yqLog(
                            '  type.format(' +
                                formatOp.index +
                                ', ' +
                                formatOp.length +
                                ', ' +
                                _fmtAttrs(formatOp.attributes) +
                                ')'
                        )
                        type.format(formatOp.index, formatOp.length, formatOp.attributes)
                    }

                    // Dump Yjs state after write
                    try {
                        const yjsDelta = type.toDelta()
                        _yqLog('  Yjs state AFTER write (' + yjsDelta.length + ' ops):')
                        yjsDelta.forEach((d, i) => {
                            if (d.insert !== undefined) {
                                _yqLog(`    [${i}] INSERT ${_snippet(d.insert)} attrs=${_fmtAttrs(d.attributes)}`)
                            } else {
                                _yqLog(`    [${i}]`, JSON.stringify(d))
                            }
                        })
                    } catch (e) {
                        _yqLog('  Could not read Yjs state:', e.message)
                    }
                })
            }
            // always check selection
            if (awareness && quillCursors) {
                const sel = quill.getSelection()
                const aw = /** @type {any} */ (awareness.getLocalState())
                if (sel === null) {
                    if (awareness.getLocalState() !== null) {
                        awareness.setLocalStateField('cursor', /** @type {any} */ (null))
                    }
                } else {
                    const anchor = Y.createRelativePositionFromTypeIndex(type, sel.index)
                    const head = Y.createRelativePositionFromTypeIndex(type, sel.index + sel.length)
                    if (
                        !aw ||
                        !aw.cursor ||
                        !Y.compareRelativePositions(anchor, aw.cursor.anchor) ||
                        !Y.compareRelativePositions(head, aw.cursor.head)
                    ) {
                        awareness.setLocalStateField('cursor', {
                            anchor,
                            head,
                        })
                    }
                }
                // update all remote cursor locations
                awareness.getStates().forEach((aw, clientId) => {
                    updateCursor(quillCursors, aw, clientId, doc, type)
                })
            }
        }
        quill.on('editor-change', this._quillObserver)
        // Pre-populate _negatedUsedFormats from the existing document content.
        // This ensures that when setContents is called, all format keys used anywhere
        // in the document are already known, so unformatted text gets explicit
        // negations (e.g. bold: false) preventing format inheritance/bleed.
        const initialDelta = type.toDelta()
        _yqLog('=== INITIAL LOAD ===')
        _yqLog('Raw Yjs delta (type.toDelta()) — ' + initialDelta.length + ' ops:')
        initialDelta.forEach((d, i) => {
            if (d.insert !== undefined) {
                _yqLog(`  [${i}] INSERT ${_snippet(d.insert)} attrs=${_fmtAttrs(d.attributes)}`)
            } else {
                _yqLog(`  [${i}]`, JSON.stringify(d))
            }
        })
        for (let i = 0; i < initialDelta.length; i++) {
            const attrs = initialDelta[i].attributes
            if (attrs) {
                for (const key in attrs) {
                    if (this._negatedUsedFormats[key] === undefined) {
                        this._negatedUsedFormats[key] = false
                    }
                }
            }
        }
        _yqLog('_negatedUsedFormats after scan:', JSON.stringify(this._negatedUsedFormats))
        // Normalize the initial delta to prevent format bleed on reload.
        // This applies the same protection as _typeObserver: explicitly set
        // all tracked formats to false on inserts that don't have them,
        // preventing Quill from inheriting formats from adjacent text.
        const normalizedDelta = initialDelta.map(d => {
            if (d.insert !== undefined) {
                const mergedAttrs = Object.assign({}, this._negatedUsedFormats, d.attributes || {})
                return Object.assign({}, d, { attributes: mergedAttrs })
            }
            return d
        })
        _yqLog('Normalized delta — ' + normalizedDelta.length + ' ops:')
        normalizedDelta.forEach((d, i) => {
            if (d.insert !== undefined) {
                _yqLog(`  [${i}] INSERT ${_snippet(d.insert)} attrs=${_fmtAttrs(d.attributes)}`)
            } else {
                _yqLog(`  [${i}]`, JSON.stringify(d))
            }
        })
        mux(() => {
            quill.setContents(normalizedDelta)
        })
        // Log what Quill actually ended up with after setContents
        try {
            const quillDelta = quill.getContents()
            _yqLog('Quill contents AFTER setContents — ' + quillDelta.ops.length + ' ops:')
            quillDelta.ops.forEach((d, i) => {
                if (d.insert !== undefined) {
                    _yqLog(`  [${i}] INSERT ${_snippet(d.insert)} attrs=${_fmtAttrs(d.attributes)}`)
                } else {
                    _yqLog(`  [${i}]`, JSON.stringify(d))
                }
            })
        } catch (e) {
            _yqLog('Could not read Quill contents after setContents:', e.message)
        }
        _yqLog('=== END INITIAL LOAD ===')
        // init remote cursors
        if (quillCursors !== null && awareness) {
            awareness.getStates().forEach((aw, clientId) => {
                updateCursor(quillCursors, aw, clientId, doc, type)
            })
            awareness.on('change', this._awarenessChange)
        }
    }
    destroy() {
        this.type.unobserve(this._typeObserver)
        this.quill.off('editor-change', this._quillObserver)
        if (this.awareness) {
            this.awareness.off('change', this._awarenessChange)
        }
    }
}
